### ISSUE TITLE
 CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.

### ISSUE THREAD

**Issue by GarreVyshnavi** [pnp/pnpjs#1231]
Hi,

I am working on node js application for MS teams app, built using yeoman generator. my application runs on ngrok. I want to use SharePoint site data in my app, so following [link](https://pnp.github.io/pnpjs/sp/#getting-started-nodejs) for fetching SharePoint data, but I am stuck with CORS error while fetching.

error shown: 
 Access to fetch at 'https://xxx.sharepoint.com/sites/xxx/_vti_bin/client.svc' from origin 'https://xxx.ngrok.io' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.

and error response as 'TypeError: Failed to fetch'

Please provide me a way to get out of this issue.

**Comment by koltyakov** [comment]
Hi @GarreVyshnavi, 

Node.js can't receive CORS restrictions error as it's a server-side process that is run in V8 on the backend but not in browser, so you're sending a request from a browser.
You can't use the mentioned authentication for a client-side in-browser application that just won't work even in theory as it doesn't suppose to.
For a client-side call, OAuth Flow should be used and [ADAL](https://pnp.github.io/pnpjs/adaljsclient/adalclient/)/MSAL. Or a request from an app should be sent to your Node.js endpoints.

**Comment by koltyakov** [comment]
@GarreVyshnavi 

The version of Node.js is not really important if you ended up using client-side auth flow. For ADAL/MSAL you don't need backend logic at all, the assets can be hosted even statically.

AdalClient helper can use SPFx's default AAD App registration to make things really simple but SPFx context is not necessarily required. You can add your own AAD App Registration, configure permissions, and use it in your app like [this](https://pnp.github.io/pnpjs/adaljsclient/adalclient/#setup-and-use-with-microsoft-graph).

For MSAL, I'd recommend reading this [article](https://spblog.net/post/2019/06/04/building-single-page-application-with-react-msal-js-and-pnpjs), ADAL is also similar in terms of AAD configuration.

**Comment by patrick-rodgers** [comment]
We have a new article on authentication coming out with the next release which might help here. You can [see it now in the source](https://github.com/pnp/pnpjs/tree/version-2/docs/authentication). CORS isn't a server issue so it seems this question is answered. Going to close, please open a new issue, link this one, and provide additional details if you continue to have issues. Thanks!

**Comment by thw0rted** [comment]
@patrick-rodgers  it looks to me like all of those articles are talking about auth via `login.microsoftonline.com`.  Do they work with on-prem?

Our SP is tied to our login AD and uses Windows auth for transparent authentication, but I'm trying to access the data from another (completely unrelated, non-Sharepoint) web application.  This works fine with CORS protections disabled in the browser (Chrome with `--disable-web-security` flag), for dev, so the concept is sound, but it sounds like the official stance is that CORS will not be supported and I have to use an OAuth flow instead.  Are there docs for OAuth with on-prem?  Should I file a new ticket for this?
